code/obj/machinery/computer/robotics.dm

```diff
@@ -22,6 +37,121 @@
 		..()
 		STOP_TRACKING
 
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if (!ui)
+			ui = new(user, src, "RoboticsControl")
+			ui.open()
+
+	ui_data(mob/user)
+		src.update_silicon_statuses()
+		src.update_ghostdrone_statuses()
+		. = list(
+			"user_is_ai" = isAI(user),
+			"user_is_cyborg" = isrobot(user),
+			"ai_names" = src.ai_names,
+			"ai_statuses" = src.ai_statuses,
+			"ai_killswitch_times" = src.ai_killswitch_times,
+			"cyborg_names" = src.cyborg_names,
+			"cyborg_statuses" = src.cyborg_statuses,
+			"cyborg_cell_charges" = src.cyborg_cell_charges,
+			"cyborg_modules" = src.cyborg_modules,
+			"cyborg_lock_times" = src.cyborg_lock_times,
+			"cyborg_killswitch_times" = src.cyborg_killswitch_times,
+			"ghostdrone_names" = src.ghostdrone_names
+		)
+
+	ui_act(action, params)
+		. = ..()
+		if (.)
+			return
+		. = TRUE
+		switch (action)
+			if ("start_ai_killswitch")
+				var/obj/item/card/id/I = usr.equipped()
+				if (istype(I))
+					if(src.check_access(I))
+						var/mob/living/silicon/ai/ai_player = by_type[/mob/living/silicon/ai][params["index"]]
```

Comment:
I feel uneasy about just using an index-accessed by_type here. Why not pass the `ref`s of them around?

Reply:
Ref needs to be passed through UI data and returned?

Reply:
You can pass a `\ref` through to `ui_data`, then pass that back in the `act`. Verify that they're a valid target still in `ui_act`, but otherwise yes, they're just strings for the purposes of `ui_data`.